In [1]:
import nltk
import re
import csv
import random
import os

# جمع ملفات التدريب والاختبار ووضعها في ملف واحد للاختبار وملف واحد للتدريب

In [2]:
# جمع ملفات التدريب في ملف واحد
file = open("train.txt", "w",encoding='utf-8')
for i in os.listdir("file/"):
    if i.endswith('.txt'):
        f="file/"+i
        fi = open(f, "r",encoding='utf-8')
        string=fi.read()
        file.write(string)
file.close()

In [3]:
# جمع ملفات الاختبار في ملف واحد
file_test = open("test.txt", "w",encoding='utf-8')
for i in os.listdir("test/"):
    if i.endswith('.txt'):
        f="test/"+i
        fi = open(f, "r",encoding='utf-8')
        string=fi.read()
        file_test.write(string)
file_test.close()

csv عدد التغريدات كبير ما فيني عدل يدويا بسبب هالشي رح عدل التغريدات ليزبط شكلن قبل ما يتخزنو بملف
pos|neg لتتوحد كل تاغات التغريدات وحدت شكل
اخدت التغريدات فقط التي بتحوي على تاغ
neg و pos حذفت الروابط لان في طلاب حاطين || قبلها وعم يروح الرابط كتاغ بدون
 tag بالنهاية ضلت التغريدات بدون روابط ومع

In [4]:
#csv عدد التغريدات كبير ما فيني عدل يدويا بسبب هالشي رح عدل التغريدات ليزبط شكلن قبل ما يتخزنو بملف 
pattern=r'pos|neg'
file= open("train.txt", "r",encoding='utf-8')
file_1= open("train_1.txt", "w",encoding='utf-8')
for i in file.readlines():
    out =i.lower()     #  pos و neg  لتتوحد كل تاغات التغريدات وحدت شكل
    if re.search(pattern,out): # اخدت التغريدات فقط التي بتحوي على تاغ
        pattern_1=r'(http\S+)|((\w(\d)?)+(.(\w(\d)?)+)?.com\S+)|(www\.[\s]+)' 
        out=re.sub(pattern_1,' ',out) #neg و pos حذفت الروابط لان في طلاب حاطين || قبلها وعم يروح الرابط كتاغ بدون
        file_1.write(out)

In [5]:
file.close()
file_1.close()

In [6]:
pattern=r'pos|neg'
file= open("test.txt", "r",encoding='utf-8')
file_1= open("test_1.txt", "w",encoding='utf-8')
for i in file.readlines():
    out =i.lower()     #  pos و neg  لتتوحد كل تاغات التغريدات وحدت شكل
    if re.search(pattern,out): # اخدت التغريدات فقط التي بتحوي على تاغ
        pattern_1=r'(http\S+)|((\w(\d)?)+(.(\w(\d)?)+)?.com\S+)|(www\.[\s]+)' 
        out=re.sub(pattern_1,' ',out) #neg و pos حذفت الروابط لان في طلاب حاطين || قبلها وعم يروح الرابط كتاغ بدون
        file_1.write(out)
file.close()
file_1.close()

train.csv تخزين التغريدات للتدريب في ملف

In [8]:
#train.csv تخزين التغريدات للتدريب في ملف
file_2= open("train_1.txt", "r",encoding='utf-8')
file_3= open("train.csv", "w",encoding='utf-8-sig')
fieldnames = ['tweet', 'tag']
writer = csv.DictWriter(file_3, fieldnames=fieldnames)
writer.writeheader()
for i in file_2.readlines():
    i = i.replace(',', ' ')
    i = i.replace('||', ',')
    file_3.write(i)
file_2.close()
file_3.close()

test.csv تخزين التغريدات للاختبار في ملف

In [9]:
#test.csv تخزين التغريدات للاختبار في ملف
file_2= open("test_1.txt", "r",encoding='utf-8')
file_3= open("test.csv", "w",encoding='utf-8-sig')
fieldnames = ['tweet', 'tag']
writer = csv.DictWriter(file_3, fieldnames=fieldnames)
writer.writeheader()
for i in file_2.readlines():
    i = i.replace(',', ' ')
    i = i.replace('||', ',')
    file_3.write(i)
file_2.close()
file_3.close()

قراءة التغريدات وتخزينها في متحول

In [6]:
# قراءة التغريدات وتخزينها في متحول
import pandas as pd
tweets=pd.read_csv('train.csv',error_bad_lines=False,names=["tweet","tag"])

In [7]:
tweets_test=pd.read_csv('test.csv',error_bad_lines=False,names=["tweet","tag"])

In [8]:
#list of tuple تحويل التغريدات ل 
data_t = [tuple(x) for x in tweets.values]

In [9]:
data_ts = [tuple(x) for x in tweets_test.values]

In [10]:
data_train=[tuple(i) for i in data_t if i[1]=="neg" or i[1]=="pos" ]

In [11]:
data_test=[tuple(i) for i in data_ts if i[1]=="neg" or i[1]=="pos" ]

In [12]:
print(len(data_train))
print(len(data_test))

9081
4727


In [13]:
data_train[10:20]

[('تحتاج الولاية الى تقنين حسب مقتضى الحال في الوكالة النظر الى عمر المرأة وسلوكها وتقنينها من ناحيه حق المرأة في إدارة اموالها وفِي حقها في مايخص ابنائها من ناحيه الحضانه واعطائها حق الولايه في حالات الطلاق وموت الزوج اذا هي مؤهله وهكذ #الشعب_يرفض_اسقاط_الولايه',
  'neg'),
 ('#الشعب_يرفض_اسقاط_الولايه اعرف من تبنى فكرة تاق اسقاط الولايه : مجنسين ابناء زوجه غير سعوديه ابناء متخلفي الحج والعمره من يعاني من مشاكل اجتماعيه من احداث السن والسفهاء الذين يسهل التغرير بهم...الخ حينها تعرف ان الموضوع يخص مجتمعات اخرى غير المجتمع السعودي ',
  'neg'),
 ('#الشعب_يرفض_اسقاط_الولايه سبحان الله الحمدالله لا إله الا الله الله أكبر سبحان الله وبحمده سبحان الله العظيم استغفرالله وأتوب إليه لا إله الا انت سبحانك اني كنت من الظالمين اللهم صل وسلم على نبينا محمد',
  'neg'),
 ('أطالب بإلغاء هذا الهاشتاق وعدم الالتفات الى مثل هذة الهاشتاقات التافهة ومن يديرها ليسوا سعوديين بل مرتزقة حمقى لايفقهون شيءً ومعظمهم لدية شهادة سادس ابتدائي أو ثالث متوسط ولَم ينهوا تعليمهم ،الشعب السعودي واعي وفاهم وأصيل لايهتم بمث

In [14]:
data_test[10:20]

[('وصف مطابق للطابور الخامس الذي يزعزع أمننا..! #إسقاط_العباية #اسقاط_الولاية #سعوديون_نرفض_دمج_المدارس  …',
  'neg'),
 ('ألقوه ألقوه لابارك الله بالضعف سكتنا عن القرارات الهمجيه حقتكم من إسقاط الولايه الى قياده المراه الى توظيف النساء ف المحلات الان ع كل أولياء الأمور التجمع امام المدرسه والاعتراض غير كذا مايفيد معهم شي نبغى تطبيق شرع الله لاأقل ولا اكثر #سعوديون_نرفض_دمج_المدارس',
  'neg'),
 ('ألقوه ألقوه لابارك الله بالضعف سمتنا عن القرارات الهمجيه حقتكم من إسقاط الولايه الو قياده المراه الى توظيف النساء ف المحلات الان ع كل أولياء الأمور التجمع انام المدرسه والاعتراض غير كذا مايفيد معهم شي #سعوديون_نرفض_دمج_المدارس',
  'neg'),
 ('#سعوديون_نرفض_دمج_المدارس 1-الغاء هيئة الامر بالمعروف والنهي عن المنكر 2-هيئه الترفيه 3- الحفلات والرقص 4-ازدياد التبرج 5-الاختلاط بالوظائف 6- الضرائب 7-السماح للنساء بقيادة السيارة 8-اسقاط الولاية 9-دمج التعليم 10- 11- 12- 13 14- 15',
  'neg'),
 ('#سعوديون_نرفض_دمج_المدا رس لن ينتهي الامر عن هذا الحد ولا هذا القرار .. كل يوم قرار جديد في تغريب المجتمع وإخر

# 2- Text Cleaning and Normalization

In [15]:
# حذف التغريدات المكررة
def del_freq (data) : 
    data_1=[t for t in (set(tuple(i) for i in data))]
    return data_1

In [16]:
# 2- Text Cleaning and Normalization
from nltk.tokenize import TreebankWordTokenizer
from nltk.corpus import stopwords
STOPWORDS = set(stopwords.words('arabic'))
def clean_stop_norm(text):
    pattern=r'@|;|؛|،|ٍٍ/|\(|\)|{|}|\[|\]|\|'
    t=re.sub(pattern,' ',str(text)) #استبدال محارف محددة
    pattern_d=r'(..)\1+'
    t = re.sub(pattern_d, r'\1', str(t))
    symbol=["ئ","ة","أ","إ","آ","ؤ"] 
    sumb_1=["ء","ه","ا","ا","ا","ء"]
    for i in range(6):
        t = str(t).replace(symbol[i], sumb_1[i])
    t = nltk.sent_tokenize(t)
    t =[w for w in t if not w in STOPWORDS]
    t = " ".join(t)
    return t

In [17]:
data_1=del_freq (data_train)

In [18]:
data_2=del_freq (data_test)

In [19]:
# عدد التغريدات بعد ازالة التكرار
print(len(data_1))
print(len(data_2))

7067
4258


In [20]:
data_train_t=[(clean_stop_norm(i[0]),i[1])for i in data_1]

In [21]:
data_test_t=[(clean_stop_norm(i[0]),i[1])for i in data_2]

In [22]:
# فصل البيانات الى اختبار وتدريب
X_train = [example[0] for example in data_train_t]
y_train = [example[1] for example in data_train_t]

X_test = [example[0] for example in data_test_t]
y_test = [example[1] for example in data_test_t]

# Bag of words

In [23]:
# Bag of words

from sklearn.feature_extraction.text import CountVectorizer

In [39]:
def bow_features(X_train, X_test):
    """
        X_train, X_val, X_test — samples        
        return Bag of words vectorized representation of each sample and vocabulary
    """
    # Create Count vectorizer with a proper parameters choice
    # Fit the vectorizer on the train set
    # Transform the train, test, and val sets and return the result
    
    count_vectorizer = CountVectorizer(ngram_range=(1, 2)) 
    X_train_Vec = count_vectorizer.fit_transform(X_train)
    X_test_Vec = count_vectorizer.transform(X_test)
    
    return X_train_Vec, X_test_Vec, count_vectorizer.vocabulary_ ,count_vectorizer

In [40]:
X_train_bag, X_test_bag, bow_vocab,count_vectorizer_bag = bow_features(X_train, X_test)

# TF-IDF

In [27]:
# TF-IDF 
from sklearn.feature_extraction.text import TfidfVectorizer

In [42]:
def tfidf_features(X_train, X_test):
    """
        X_train, X_val, X_test — samples        
        return TF-IDF vectorized representation of each sample and vocabulary
    """
    # Create TF-IDF vectorizer with a proper parameters choice
    # Fit the vectorizer on the train set
    # Transform the train, test, and val sets and return the result
    
    
    tfidf_vectorizer = TfidfVectorizer(min_df=2, max_df=0.7, ngram_range=(1, 2))
    X_train = tfidf_vectorizer.fit_transform(X_train)
    X_test = tfidf_vectorizer.transform(X_test)
    
    return X_train, X_test, tfidf_vectorizer.vocabulary_,tfidf_vectorizer

In [43]:
X_train_tfidf, X_test_tfidf, tfidf_vocab ,tfidf_vectorizer_tf= tfidf_features(X_train, X_test)

In [30]:
# Classify
from sklearn.multiclass import OneVsRestClassifier
from sklearn.linear_model import LogisticRegression, RidgeClassifier
def train_classifier(X_train, y_train):
    classifier = RidgeClassifier()
    classifier.fit(X_train, y_train)
    return classifier

In [31]:
classifier_bag = train_classifier(X_train_bag, y_train)
classifier_tfidf = train_classifier(X_train_tfidf, y_train)

In [32]:
y_test_predicted_labels_bag = classifier_bag.predict(X_test_bag)
y_test_predicted_scores_bag = classifier_bag.decision_function(X_test_bag)

y_test_predicted_labels_tfidf = classifier_tfidf.predict(X_test_tfidf)
y_test_predicted_scores_tfidf = classifier_tfidf.decision_function(X_test_tfidf)

In [33]:
for i in range(3):
    print('Tweet:\t{}\nTrue Tag:\t{}\nPredicted label BoW:\t{}\nPredicted label Tf_Idf:\t{}\n\n'.format(
        X_test[i][:100] + "...",y_test[i],y_test_predicted_labels_bag[i], y_test_predicted_labels_tfidf[i])
    )

Tweet:	اغلب النساء لم يكن الرادع لهن من السفر بدون محرم قوانين حكوميه! بل كان الرادع الرءيسي لديهن طاعه الل...
True Tag:	neg
Predicted label BoW:	neg
Predicted label Tf_Idf:	neg


Tweet:	#السعوديه ... قريبا ستذهب نساءنا للحج بدون ذي محرم كما تراجعوا عن اخذ جواز سفر المراه من دون اذن ولي...
True Tag:	neg
Predicted label BoW:	neg
Predicted label Tf_Idf:	neg


Tweet:	الحملات المغرضه تتغاضى عن النماذج المتميزه لغالبيهالاباء والازواج والاخوهالذين يرعون ويهتمون بامهاته...
True Tag:	neg
Predicted label BoW:	pos
Predicted label Tf_Idf:	pos




In [34]:
from sklearn.metrics import accuracy_score

print('Bag-of-words Accuracy: '+ str(accuracy_score(y_test, y_test_predicted_labels_bag)))
print('Tfidf Accuracy: ' + str(accuracy_score(y_test, y_test_predicted_labels_tfidf)))


Bag-of-words Accuracy: 0.7909816815406294
Tfidf Accuracy: 0.8048379520901832


وصلتلو Accuracy اعلا 
بتعديل min ,max
Bag-of-words Accuracy: 0.791451385627055
Tfidf Accuracy: 0.8046031000469704

# 3-لتصنيف تغريدة يقوم المستخدم بإدخالها

In [57]:
# 3-لتصنيف تغريدة يقوم المستخدم بإدخالها
tweet=input("enter tweet ")

print(classifier_bag.predict(count_vectorizer_bag.transform([tweet])[0]))
print(classifier_tfidf.predict(tfidf_vectorizer_tf.transform([tweet])[0]))

enter tweet رفع الولاية
['pos']
['pos']


# 4-stemmingالتصنيف بعد القيام بعمليات

In [38]:
# 4-stemmingالتصنيف بعد القيام بعمليات   
porter = nltk.PorterStemmer()

def stem_text(in_text):
    out_text = " ".join(porter.stem(w) for w in in_text.split())
    return out_text

In [39]:
data_train_stem=[(stem_text(i[0]),i[1])for i in data_train_t]
data_test_stem=[(stem_text(i[0]),i[1])for i in data_test_t]

In [40]:
X_train = [example[0] for example in data_train_stem]
y_train = [example[1] for example in data_train_stem]

X_test = [example[0] for example in data_test_stem]
y_test = [example[1] for example in data_test_stem]

In [41]:
X_train_bag, X_test_bag, bow_vocab = bow_features(X_train, X_test)

In [42]:
X_train_tfidf, X_test_tfidf, tfidf_vocab = tfidf_features(X_train, X_test)

In [43]:
classifier_bag = train_classifier(X_train_bag, y_train)
classifier_tfidf = train_classifier(X_train_tfidf, y_train)

In [44]:
y_test_predicted_labels_bag = classifier_bag.predict(X_test_bag)
y_test_predicted_scores_bag = classifier_bag.decision_function(X_test_bag)

y_test_predicted_labels_tfidf = classifier_tfidf.predict(X_test_tfidf)
y_test_predicted_scores_tfidf = classifier_tfidf.decision_function(X_test_tfidf)

In [58]:
for i in range(3):
    print('Tweet:\t{}\nTrue Tag:\t{}\nPredicted label BoW:\t{}\nPredicted label Tf_Idf:\t{}\n\n'.format(
        X_test[i][:100] + "...",y_test[i],y_test_predicted_labels_bag[i], y_test_predicted_labels_tfidf[i])
    )

Tweet:	سعوديات نطلب اسقاط الولايه1000 هذا التاق بدون قاءده تغرد فيه كل امراه سعوديه حتى لو لم يتجاوز متابعي...
True Tag:	pos
Predicted label BoW:	pos
Predicted label Tf_Idf:	pos


Tweet:	الشعب يرفض اسقاط الولايه قال صلى الله عليه وسلم"كُلُّكُمْ رَاعٍ وَكُلُّكُمْ مَسْءول عَنْ رَعِيَّتِهِ...
True Tag:	neg
Predicted label BoW:	neg
Predicted label Tf_Idf:	neg


Tweet:	اسقاط الولايه بعد الـ 18 قد يسقط حق الميراث لمن تسقط ولايته والزواج سيعيد الفتاه لولايه الزوج قرار ا...
True Tag:	pos
Predicted label BoW:	neg
Predicted label Tf_Idf:	neg




In [46]:
print('Bag-of-words Accuracy: '+ str(accuracy_score(y_test, y_test_predicted_labels_bag)))
print('Tfidf Accuracy: ' + str(accuracy_score(y_test, y_test_predicted_labels_tfidf)))

Bag-of-words Accuracy: 0.7912165335838421
Tfidf Accuracy: 0.8046031000469704


ما اثر steeming على التدريب
Bag-of-words Accuracy: 0.7912165335838421
Tfidf Accuracy: 0.8046031000469704

# 5- التصنيف بعد حذف الهاشتاغ

In [48]:
# 5- التصنيف بعد حذف الهاشتاغ
pattern=r'#(\S+_\S+)+'

def del_hash_tag(text):
    out=re.sub(pattern,' ',text)
    return out

In [49]:
data_train_has=[(del_hash_tag(i[0]),i[1])for i in data_train_t]
data_test_has=[(del_hash_tag(i[0]),i[1])for i in data_test_t]

X_train = [example[0] for example in data_train_has]
y_train = [example[1] for example in data_train_has]

X_test = [example[0] for example in data_test_has]
y_test = [example[1] for example in data_test_has]

X_train_bag, X_test_bag, bow_vocab = bow_features(X_train, X_test)
X_train_tfidf, X_test_tfidf, tfidf_vocab = tfidf_features(X_train, X_test)
classifier_bag = train_classifier(X_train_bag, y_train)
classifier_tfidf = train_classifier(X_train_tfidf, y_train)
y_test_predicted_labels_bag = classifier_bag.predict(X_test_bag)
y_test_predicted_scores_bag = classifier_bag.decision_function(X_test_bag)

y_test_predicted_labels_tfidf = classifier_tfidf.predict(X_test_tfidf)
y_test_predicted_scores_tfidf = classifier_tfidf.decision_function(X_test_tfidf)

In [59]:
for i in range(3):
    print('Tweet:\t{}\nTrue Tag:\t{}\nPredicted label BoW:\t{}\nPredicted label Tf_Idf:\t{}\n\n'.format(
        X_test[i][:100] + "...",y_test[i],y_test_predicted_labels_bag[i], y_test_predicted_labels_tfidf[i])
    )

Tweet:	سعوديات نطلب اسقاط الولايه1000 هذا التاق بدون قاءده تغرد فيه كل امراه سعوديه حتى لو لم يتجاوز متابعي...
True Tag:	pos
Predicted label BoW:	pos
Predicted label Tf_Idf:	pos


Tweet:	الشعب يرفض اسقاط الولايه قال صلى الله عليه وسلم"كُلُّكُمْ رَاعٍ وَكُلُّكُمْ مَسْءول عَنْ رَعِيَّتِهِ...
True Tag:	neg
Predicted label BoW:	neg
Predicted label Tf_Idf:	neg


Tweet:	اسقاط الولايه بعد الـ 18 قد يسقط حق الميراث لمن تسقط ولايته والزواج سيعيد الفتاه لولايه الزوج قرار ا...
True Tag:	pos
Predicted label BoW:	neg
Predicted label Tf_Idf:	neg




In [51]:
print('Bag-of-words Accuracy: '+ str(accuracy_score(y_test, y_test_predicted_labels_bag)))
print('Tfidf Accuracy: ' + str(accuracy_score(y_test, y_test_predicted_labels_tfidf)))

Bag-of-words Accuracy: 0.7806481916392672
Tfidf Accuracy: 0.7881634570220761


 انخفضت بعد حذف الهاشتاغ
Bag-of-words Accuracy: 0.7806481916392672
Tfidf Accuracy: 0.7881634570220761

# 6-  التصنيف بعد تحويل الهاشتاغات

In [52]:
# 6-  التصنيف بعد تحويل الهاشتاغات

def hash_tag(text):
    out=re.sub('#','',text)
    out=re.sub('_',' ',out)
    return out

In [53]:
data_train_has1=[(hash_tag(i[0]),i[1])for i in data_train_t]
data_test_has1=[(hash_tag(i[0]),i[1])for i in data_test_t]

X_train = [example[0] for example in data_train_has1]
y_train = [example[1] for example in data_train_has1]

X_test = [example[0] for example in data_test_has1]
y_test = [example[1] for example in data_test_has1]

X_train_bag, X_test_bag, bow_vocab = bow_features(X_train, X_test)
X_train_tfidf, X_test_tfidf, tfidf_vocab = tfidf_features(X_train, X_test)
classifier_bag = train_classifier(X_train_bag, y_train)
classifier_tfidf = train_classifier(X_train_tfidf, y_train)
y_test_predicted_labels_bag = classifier_bag.predict(X_test_bag)
y_test_predicted_scores_bag = classifier_bag.decision_function(X_test_bag)

y_test_predicted_labels_tfidf = classifier_tfidf.predict(X_test_tfidf)
y_test_predicted_scores_tfidf = classifier_tfidf.decision_function(X_test_tfidf)

In [60]:
for i in range(3):
    print('Tweet:\t{}\nTrue Tag:\t{}\nPredicted label BoW:\t{}\nPredicted label Tf_Idf:\t{}\n\n'.format(
        X_test[i][:100] + "...",y_test[i],y_test_predicted_labels_bag[i], y_test_predicted_labels_tfidf[i])
    )

Tweet:	سعوديات نطلب اسقاط الولايه1000 هذا التاق بدون قاءده تغرد فيه كل امراه سعوديه حتى لو لم يتجاوز متابعي...
True Tag:	pos
Predicted label BoW:	pos
Predicted label Tf_Idf:	pos


Tweet:	الشعب يرفض اسقاط الولايه قال صلى الله عليه وسلم"كُلُّكُمْ رَاعٍ وَكُلُّكُمْ مَسْءول عَنْ رَعِيَّتِهِ...
True Tag:	neg
Predicted label BoW:	neg
Predicted label Tf_Idf:	neg


Tweet:	اسقاط الولايه بعد الـ 18 قد يسقط حق الميراث لمن تسقط ولايته والزواج سيعيد الفتاه لولايه الزوج قرار ا...
True Tag:	pos
Predicted label BoW:	neg
Predicted label Tf_Idf:	neg




In [55]:
print('Bag-of-words Accuracy: '+ str(accuracy_score(y_test, y_test_predicted_labels_bag)))
print('Tfidf Accuracy: ' + str(accuracy_score(y_test, y_test_predicted_labels_tfidf)))

Bag-of-words Accuracy: 0.8062470643494598
Tfidf Accuracy: 0.8163457022076092


 ارتفعت بعد فصل الهاشتاغ
 Bag-of-words Accuracy: 0.8062470643494598
 Tfidf Accuracy: 0.8163457022076092